design-proposals/klip-33-key-format.md

```diff
@@ -0,0 +1,461 @@
+
+`KEY_FORMAT` will not be a required property _if_ `ksql.persistence.default.format.key` is set.
+
+Providing `FORMAT` will set both the key and value formats. Providing `FORMAT` along with either 
+`KEY_FORMAT` or `VALUE_FORMAT` will result in an error.
+
+### Server configs
+
+The following new configuration options will be added. These configurations can be set globally, 
+within the application property file, or locally, via the `SET` command.
+
+* `ksql.persistence.default.format.key`: the default key format.
+
+## Design
+
+The new `KEY_FORMAT` or `FORMAT` property will be supported where ever the current `VALUE_FORMAT` is
+supported. Namely:
+
+ * In `CRREATE STREAM` and `CREATE TABLE` statements.
+ * In `CREATE STREAM AS SELECT` and `CREATE TABLE AS SELECT` statements.
+ 
+The key format will follow the same inheritance rules as the current value format. Namely: any 
+derived stream will inherit the format of its leftmost source, unless the format is explicitly set
+in the with clause.
+
+For example:
+
+```sql
+-- Creates a table over a changelog topic with AVRO key and JSON value:
+CREATE TABLE USERS (
+    ID BIGINT PRIMARY KEY,
+    NAME STRING
+  ) WITH (
+    KAFKA_TOPIC='USERS',
+    KEY_FORMAT='AVRO',
+    VALUE_FORMAT='JSON'
+  );
+
+-- Creates a stream over a topic with JSON key and value:
+CREATE STREAM BIDS (
+    ITEM_ID BIGINT KEY,
+    USER_ID BIGINT,
+    AMMOUNT INT
+  ) WITH (
+    KAFKA_TOPIC='USERS',
+    FORMAT='JSON'
+  );
+
+-- Change the key format of a stream:
+CREATE STREAM AVRP_BIDS 
+  WITH WITH (
+    KEY_FORMAT='AVRO'
+  ) AS
+    SELECT * FROM BIDS;
+
+-- Creates an enriched stream. The key format is inherited from the leftmost source, i.e. JSON:
+CREATE STREAM ENRICHED_BIDS AS 
+  SELECT *
+  FROM BIDS 
+   JOIN USERS ON BIDS.USER_ID = USERS.ID;
+```
+
+For formats that support integration with the schema registry, the key schema will be read and 
+registered with the Schema Registry as needed, following the same pattern as the value schema in 
+the current product. 
+
+If a `CREATE TABLE` or `CREATE STREAM` statement does not include a `KEY_FORMAT` property, the 
+key format is picked up from the `ksql.persistence.default.format.key` system configuration. If this
+is not set, then an error is returned.  Note: The server config will have this set to `KAFKA` to 
+maintain backwards compatibility with current system by default.
+
+### Implementation
+
+The system already serializes the key format of source, intermediate and sink topics as part of the
+query plan, meaning it should be fairly easily to plug in new formats. 
+
+Validation will be added to ensure only supported key formats are set, and that key column data types
+are supported by key formats.
+
+Most existing functionality should _just work_, as the key format only comes into play during 
+(de)serialization, (obviously). The only area where additional work is expected are joins, single
+key wrapping and key-less streams.
+
+#### Joins
```

Comment:
Sorry I am late to the party. 
What happens in this scenario: ```CREATE TABLE test as SELECT * from test1 join test2``` where the `test1` join key has format `f1` and `test2` join key has format` f2` . The output key format is inherited from the left-most source hence will be `f1`? If the algorithm chooses that the cheapest partitioning strategy is to partition by `f2`, does this mean that we need an extra partitioning at the end for the sink topic to be in `f1` format?

If yes, then this should be part of the decision choosing which side to repartition, no?
So the heuristics for the algorithm should take into account:
1) What is the output format? 
2) What keys are the sources partitioned by? 

And the definition of cheapest should take into account the cost of partitioning at the end to match the final output format as well as the cost of re-partitioning the inputs.

Reply:
In addition to this, will this be supported 
``` 
CREATE TABLE TEST 
  WITH (
    KEY_FORMAT='F3'
  ) AS
    SELECT * FROM TEST1 JOIN TEST2;
```
If yes, then the output format can differ completely from the input formats which will definitely necessitate a final repartition for the sink topic.

Reply:
> which will definitely necessitate a final repartition for the sink topic.

Well, yes, but this is not really an issue as the "repartitioning" is for free: we write the data into the sink topic anyway and the costs is the same with or without changing the partitioning.

Reply:
How is the cost the same? If the partitioning changes, isn't there network overhead and data shuffling involved?

Reply:
There is not more network overhead involved. Every message is written once.

Note, in general when we talk about repartitioning overhead, the issue is that we need to write the data and read it back to shuffle records. (And with no repartitioning, we don't neither write nor read to a topic, ie, the "overhead/cost" of repartitioning is one write and one read operation per record.) -- However, when we write to a sink anyway, we don't have more write operations if we change the partitioning.

Maybe, the only "overhead" we might have is that the producer may need to open more network connections as it may need to write data into different partitions that might be hosted on different brokers. However, this seems negligible.